crates/log_probe/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+use composer_api::{self, Client, LogStats};
+use log::{self, Level};
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        mpsc::{sync_channel, Receiver, SyncSender},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum LogProbeError {
+    #[error("Network error: {0}")]
+    NetworkError(String),
+}
+
+pub struct LogProbe {
+    tx: SyncSender<AggregatorMessage>,
+    shutdown: Arc<AtomicBool>,
+}
+
+impl Drop for LogProbe {
+    fn drop(&mut self) {
+        self.shutdown.store(true, Ordering::Relaxed);
+    }
+}
+
+enum AggregatorMessage {
+    AddRecord(Level),
+    Tick,
+}
+
+impl LogProbe {
+    pub fn new(
+        server_address: Option<String>,
+        report_frequency: Duration,
+    ) -> Result<Self, LogProbeError> {
+        let shutdown = Arc::<AtomicBool>::default();
+        let client = if let Some(address) = server_address {
+            Client::new(address)
+        } else {
+            Client::try_default()
+        }
+        .map_err(|e| LogProbeError::NetworkError(format!("{e}")))?;
+
+        let (tx, rx) = sync_channel::<AggregatorMessage>(100);
+        spawn_aggregator_thread(rx, client, shutdown.clone());
+        spawn_tick_thread(tx.clone(), shutdown.clone(), report_frequency);
+
+        Ok(Self { tx, shutdown })
+    }
+}
+
+fn spawn_tick_thread(
+    tx: SyncSender<AggregatorMessage>,
+    shutdown: Arc<AtomicBool>,
+    report_frequency: Duration,
+) {
+    std::thread::spawn(move || {
+        let start = Instant::now();
+        for deadline in (0..).map(|i| start + i * report_frequency) {
+            if let Err(e) = tx.send(AggregatorMessage::Tick) {
+                eprintln!("Failed to communicate with aggregator thread: {e}")
+            }
+
+            if shutdown.load(Ordering::Relaxed) {
+                break;
+            }
+
+            std::thread::sleep(deadline - Instant::now());
+        }
```

Comment:
If I understand this right, the `Instant` juggling is a neat trick to ensure that over time we don't drift away from the given `report_frequency`, right?

I may be overthinking this, but in case we fall behind by more than the duration of one period we'll then catch up by ticking really fast. Don't we rather want is to skip the completely missed iterations and continue with the given frequency?

I.e. something like

```suggestion
        let mut last_iteration = Instant::now();
        loop {
            if let Err(e) = tx.send(AggregatorMessage::Tick) {
                eprintln!("Failed to communicate with aggregator thread: {e}")
            }

            if shutdown.load(Ordering::Relaxed) {
                break;
            }
    
            let time_left = last_iteration + report_frequency - Instant::now();
            last_iteration = Instant::now();
            std::thread::sleep(time_left);
        }
```

Reply:
Hmm you're right, this function wouldn't behave too nicely if the channel gets clogged up 😅.

There's something I did wrong, which is using `send` instead of `try_send`. My intention was to use a nonblocking call so ticks could be missed in case the channel was full, which I think would achieve a similar behaviour to what you're suggesting.

Are you OK with moving with `send` to `try_send`? If we do that, there shouldn't be anything stopping this thread from sticking close to the desired frequency/period.

Reply:
Yeah, definitely! Go ahead. .)